### PR TITLE
feat(oauth2): add coros provider to list of clientProviders

### PR DESCRIPTION
**Pull request by hassan254-prog**
## Describe your changes

- Add `coros` and `coros-sandbox` to the list of `shouldUseProviderClient()`. The refresh of tokens in `coros` is a little bit different; when the token is refreshed, the only response received is:
```
{
"result": "0000",
"message": "OK"
}
```
We need to let Nango know when the token will expire.
## Issue ticket number and link
[EXT-101](https://linear.app/nango/issue/EXT-101/add-support-for-coros-api)

This will build on top of [2449](https://github.com/NangoHQ/nango/pull/2449)

### CODE REVIEW

File: packages/shared/lib/clients/provider.client.ts [hunk 1]
```diff
@@ -373,6 +388,66 @@ class ProviderClient {
...
+            grant_type: 'authorization_code'
+        };
+        try {
+            const response = await axios.post(tokenUrl, body, { headers: headers });
+            if (response.status === 200 && response.data !== null) {
+                return {
+                    access_token: response.data['access_token'],
+                    refresh_token: response.data['refresh_token'],
+                    expires_in: response.data['expires_in'],
+                    openId: response.data['openId']
```
Comment (TBonnin) [inline packages/shared/lib/clients/provider.client.ts:409]: should we make sure those attributes actually exists? maybe using zod validation? I know we are not doing it for the other ones but it is never too late to start. :D

File: packages/shared/lib/clients/provider.client.ts [hunk 2]
```diff
@@ -373,6 +388,66 @@ class ProviderClient {
...
+
+        const body = {
+            client_id: clientId,
+            refresh_token: refreshToken,
+            client_secret: clientSecret,
+            grant_type: 'refresh_token'
+        };
+        try {
+            const response = await axios.post(refreshTokenUrl, body, { headers: headers });
+            if (response.status === 200 && response.data !== null && response.data.result === '0000') {
```
Comment (TBonnin) [inline packages/shared/lib/clients/provider.client.ts:442]: can we add a comment to describe what's coros api response in case of refresh success?

**Review by TBonnin — APPROVED** [review]

File: packages/shared/lib/clients/provider.client.ts (hunk 1 above)
Comment (bodinsamuel) [reply packages/shared/lib/clients/provider.client.ts:409]: This file needs a full a refacto tbh not sure it's worth starting an initiative

File: packages/shared/lib/clients/provider.client.ts [hunk 3]
```diff
@@ -373,6 +388,66 @@ class ProviderClient {
...
+                return {
+                    access_token: accessToken,
+                    refresh_token: refreshToken,
+                    expires_in: corosExpiresIn
+                };
+            }
+            throw new NangoError('coros_token_refresh_request_error');
```
Comment (bodinsamuel) [inline packages/shared/lib/clients/provider.client.ts]: we only have one error for refresh `refresh_token_external_error` + payload (should be json)

**Review by bodinsamuel — COMMENTED** [review]
Minor comments otherwise lgtm

File: packages/shared/lib/clients/provider.client.ts (hunk 3 above)
Comment (hassan254-prog) [reply packages/shared/lib/clients/provider.client.ts]: Thank you for pointing this out 😊!

File: packages/shared/lib/clients/provider.client.ts (hunk 2 above)
Comment (hassan254-prog) [reply packages/shared/lib/clients/provider.client.ts:442]: @TBonnin , I agree 💯, just [updated](https://github.com/NangoHQ/nango/pull/2476/commits/afbe4fd40e82d562f16f8c13f77188b51a75df1a) my PR. Thanks!

File: packages/shared/lib/clients/provider.client.ts [hunk 4]
```diff
@@ -373,6 +388,72 @@ class ProviderClient {
...
+                    access_token: response.data['access_token'],
+                    refresh_token: response.data['refresh_token'],
+                    expires_in: response.data['expires_in'],
+                    openId: response.data['openId']
+                };
+            }
+
+            throw new NangoError('coros_token_request_error');
+        } catch (e: any) {
+            throw new NangoError('coros_token_request_error', e.message);
```
Comment (bodinsamuel) [inline packages/shared/lib/clients/provider.client.ts]: same message about refresh_token_external_error

Comment (hassan254-prog) [reply packages/shared/lib/clients/provider.client.ts]: @bodinsamuel, wouldn't it be better if we added `request_token_external_error` to the list of NangoError `types`?

Comment (bodinsamuel) [reply packages/shared/lib/clients/provider.client.ts]: ah yes I didn't saw it wasn't a refresh 👌🏻

Comment (hassan254-prog) [reply packages/shared/lib/clients/provider.client.ts]: Thanks for clarifying this 🙏. Updated in [here](https://github.com/NangoHQ/nango/pull/2476/commits/c37778ebe37e04d4938eba0465885f4e4c47a4be).

**Review by bodinsamuel — APPROVED** [review]

**Pull request merged**